BaseTools/Source/C/GenFw/GenFw.c

```diff
@@ -443,6 +443,58 @@ Routine Description:
   return STATUS_SUCCESS;
 }
 
+// MU_CHANGE begin
+BOOLEAN
+IsNxCompatCompliant (
+  EFI_IMAGE_OPTIONAL_HEADER_UNION  *PeHdr
+  )
+/*++
+
+Routine Description:
+
+  Checks if the Pe image is nxcompat. i.e. section alignment >= 4k, not Write/Execute
+
+Arguments:
+
+  PeHdr      The Pe header
+
+Returns:
+
+  TRUE       The PE is nx compat compliant
+  FALSE      The PE is not nx compat compliant
+--*/
+{
+  EFI_IMAGE_SECTION_HEADER     *SectionHeader;
+  EFI_IMAGE_OPTIONAL_HEADER64  *Optional64;
+  UINT32                       Index;
+  UINT32                       Mask;
+
+  // Verify PE is 64 bit
+  if (!(PeHdr->Pe32.OptionalHeader.Magic == EFI_IMAGE_NT_OPTIONAL_HDR64_MAGIC)) {
+    return FALSE;
+  }
+
+  // Verify Section Alignment is divisible by 4K
+  Optional64 = (EFI_IMAGE_OPTIONAL_HEADER64 *)&PeHdr->Pe32.OptionalHeader;

```

Comment:
https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#optional-header-image-only

I think the optional header needs to be verified that it exists before we can reference it

Reply:
I think we are able to assume the option header exists as we are working with image files only (.dll) and they are required for image files... but Better safe then sorry :) I've added the check